torch/csrc/cuda/Module.cpp

```diff
@@ -519,32 +519,110 @@ PyObject* THCPModule_resetPeakMemoryStats(PyObject* _unused, PyObject* arg) {
   Py_RETURN_NONE;
 }
 
+struct Frame {
+  PyCodeObject* code;
+  int lasti;
+};
+
+struct StackContext : public c10::cuda::CUDACachingAllocator::Context {
+  std::vector<Frame> frames;
+  ~StackContext() {
+    for (auto& f : frames) {
+      Py_XDECREF((PyObject*)f.code);
+    }
+  }
+  static std::unique_ptr<c10::cuda::CUDACachingAllocator::Context> gather() {
+    py::gil_scoped_acquire acquire;
+    auto r = std::make_unique<StackContext>();
+    PyFrameObject* f = PyEval_GetFrame();
+    Py_XINCREF(f);
+    while (f) {
+      r->frames.emplace_back(Frame{PyFrame_GetCode(f), PyFrame_GetLasti(f)});
+      auto f_back = PyFrame_GetBack(f);
+      Py_XDECREF(f);
+      f = f_back;
+    }
+    return r;
+  }
+};
+
 PyObject* THCPModule_memorySnapshot(PyObject* _unused, PyObject* noargs) {
   HANDLE_TH_ERRORS
 
   using c10::cuda::CUDACachingAllocator::BlockInfo;
+  using c10::cuda::CUDACachingAllocator::History;
   using c10::cuda::CUDACachingAllocator::SegmentInfo;
 
-  const auto segmentInfoToDict = [](const SegmentInfo& segmentInfo) {
+  py::str device_s = "device";
```

Comment:
nit you could have bound these to python in a private namespace to be able to access them from the python code

Reply:
I wanted to avoid having to do static initialization of them. The previous version just passed strings to the pybind11 API, but they would end up creating a lot of string copies of the keys, which would then create large pickle archives when saving a memory dump. This was just the simplest fix for that.